src/sdk/pynni/nni/compression/torch/utils/sensitivity_analysis.py

```diff
@@ -0,0 +1,210 @@
+            [0.25, 0.5, 0.75] for an example, the SensitivityAnalysis will prune 25% 50% 75%
+            weights gradually for each layer.
+
+        prune_type : str
+            The pruner type used to prune the conv layers, default is 'l1',
+            and 'l2', 'fine-grained' is also supported.
+        early_stop : float
+            If this flag is set, the sensitivity analysis
+            for a conv layer will early stop when the accuracy
+            drop already reach the value of early_stop (0.05 for example).
+            The default value is 1.0, which means the analysis won't stop
+            until all given sparsities are tested.
+
+        """
+        self.model = model
+        self.val_func = val_func
+        self.target_layer = OrderedDict()
+        self.ori_state_dict = copy.deepcopy(self.model.state_dict())
+        self.target_layer = {}
+        self.sensitivities = {}
+        if sparsities is not None:
+            self.sparsities = sorted(sparsities)
+        else:
+            self.sparsities = np.arange(0.1, 1.0, 0.1)
+        self.sparsities = [np.round(x, 2) for x in self.sparsities]
+        self.Pruner = L1FilterPruner
+        if prune_type == 'l2':
+            self.Pruner = L2FilterPruner
+        elif prune_type == 'fine-grained':
+            self.Pruner = LevelPruner
+        self.early_stop = early_stop
+        self.ori_acc = None  # original accuracy for the model
+        # already_pruned is for the iterative sensitivity analysis
+        # For example, sensitivity_pruner iteratively prune the target
+        # model according to the sensitivity. After each round of
+        # pruning, the sensitivity_pruner will test the new sensitivity
+        # for each layer
+        self.already_pruned = {}
+        self.model_parse()
+
+    @property
+    def layers_count(self):
+        return len(self.target_layer)
+
+    def model_parse(self):
+        for name, submodel in self.model.named_modules():
+            for op_type in SUPPORTED_OP_TYPE:
+                if isinstance(submodel, op_type):
+                    self.target_layer[name] = submodel
+                    self.already_pruned[name] = 0
+
+    def analysis(self, val_args=None, val_kwargs=None, start=0, end=None):
+        """
+        This function analyze the sensitivity to pruning for
+        each conv layer in the target model.
+        If start and end are not set, we analyze all the conv
+        layers by default. Users can specify several layers to
+        analyze or parallelize the analysis process easily through
+        the start and end parameter.
+
+        Parameters
+        ----------
+        val_args : list
+            args for the val_function
+        val_kwargs : dict
+            kwargs for the val_funtion
+        start : int
+            Layer index of the sensitivity analysis start.
+        end : int
+            Layer index of the sensitivity analysis end.
```

Comment:
I think you should define the meaning of index. How users know index of each layer?

Reply:
I agree that leave the parallel options to users may confuse them, How about we not expose these options and parallelize the analysis automatically according to the available GPUs? We can have a quick discussion about this point.